Features rajouter des ennemis.

luacsoulet opened the pull request.

luacsoulet committed:
Add enemies data

luacsoulet committed:
update gameSprites

luacsoulet committed:
Create Enemy type data

luacsoulet committed:
Update WeaponComponent

- Update WeaponComponent to handle directional shooting
- Add support for different bullet textures
- Make player bullets follow ship rotation
- Make enemy bullets move downward
- Fix saucer animation to play shooting sequence before firing

luacsoulet committed:
Implement specialized shooting bullets for scripted data

- Remove idle animation for saucer enemies
- Ensure other enemy types maintain their idle animations

luacsoulet committed:
Update Player.ts

luacsoulet committed:
Enhance enemy system with new types

- Add support for multiple enemy types (swarm, saucer, shooter)
- Implement specialized enemy spawning functions for each type
- Configure enemy-specific properties via JSON data
Improve bullet system with sprite-based projectiles

luacsoulet committed:
Create formation json for différent shapes

luacsoulet committed:
Create FormationData.ts

luacsoulet committed:
Implement enemy formations

- Add FormationManager to handle different enemy formations (line, circle, triangle, square, zigzag)
- Configure formations via JSON data with customizable parameters
- Fix enemy shooting animations for saucer type enemies
- Improve bullet system with directional shooting
- Make player bullets follow ship rotation
- Make enemy bullets move downward
- Fix physics body activation with proper TypeScript typing
- Add debug logging for formation spawning
- Fix enemy type initialization to ensure correct enemy types spawn
- Fix angle setting on enemy bodies (apply to entity instead of physics body)
- Balance enemy spawn rates and formation timing

luacsoulet committed:
Add FormationManager component for enemy spawning patterns

- Implement multiple formation types (line, circle, triangle, square, zigzag)
- Add support for random formation generation
- Configure enemy placement based on formation parameters
- Support different enemy types within formations
- Add proper error handling for missing formation data
- Implement geometric placement algorithms for each formation type
- Add debug logging for formation spawning

luacsoulet committed:
Enhance Enemy class with type-specific behaviors

rthery commented on public/assets/Data/enemies.json:
```diff
@@ -0,0 +1,95 @@
+{
```
Tres bien la definition des enemis via un JSON

rthery commented on public/assets/Data/enemies.json:
```diff
@@ -0,0 +1,95 @@
+        "points": 1,
+        "textures": {
+            "default": "ufoRed",
+            "speed_1": "ufoRed-shoot2",
```
Un peu bizarre de nomme speed_x et d'associer une texture de shoot

rthery commented on src/components/FormationManager.ts:
```diff
@@ -0,0 +1,171 @@
+    private createEnemy(x: number, y: number, type: string): Enemy {
+        const enemy = this.enemies.get() as Enemy;
+        if (enemy) {
+            enemy.init(this.enemyBullets, type);
```
Attention l'init se fait deja quand le groupe crée ses enfants, du coup il sera appelé deux fois, voir plus que ca une fois que le groupe les recycles

On essaye generalement d'avoir l'init appeler qu'une seule fois dans le cycle de vie d'une instance

rthery commented on src/components/FormationManager.ts:
```diff
@@ -0,0 +1,171 @@
+import { Enemy } from "../entities/Enemy";
```
Bonne initiative d'avoir deleguer la gestion de ton spawn et formation via cette classe

rthery commented on src/components/FormationManager.ts:
```diff
@@ -0,0 +1,171 @@
+
+    private spawnCircleFormation(x: number, y: number, count: number, type: string, radius: number) {
+        const circle = new Phaser.Geom.Circle(x, y, radius);
+        const points = circle.getPoints(count);
```
Bonne utilisation de l'API de Phaser

rthery commented on src/components/WeaponComponent.ts:
```diff
@@ -4,33 +4,61 @@ import { Bullet } from "../entities/Bullet";
+            let velocityX = 0;
+            let velocityY = 0;
+
+            if (isPlayer) {
```
Il faudrait eviter que ton component Weapon est une partie de logique specifique a une entite
Tu pourrais modifier shoot, pour optionnellement permettre la logique dans ce if, et ensuite permettre de l'activier depuis ta classe Player

rthery commented on src/components/WeaponComponent.ts:
```diff
@@ -4,33 +4,61 @@ import { Bullet } from "../entities/Bullet";
+                velocityX,
+                velocityY
+            );
+            (bullet as any).damage = this.bulletDamage;
```
Pourquoi ne pas ajouter une property damage sur bullet plutot que de le cast en `any` ?

rthery commented on src/entities/Enemy.ts:
```diff
@@ -2,67 +2,232 @@ import { Entity } from "./Entity";
+        this.setupComponents(bulletsGroup);
 
-        // Configurer le corps circulaire
+        this.setupAnimations();
```
Tres bien de breakdown cette grosse methode en plusieurs sous methode

rthery commented on src/gameData/EnemyData.ts:
```diff
@@ -0,0 +1,22 @@
+type EnemyConfig = {
```
Attention le nom du type ne correspond pas au nom de ton fichier

rthery reviewed (commented):
Attention aux problemes potentiel des calls sur l'init multiple, mais globalement ton choix de classes et ton code est tres propre, bravo!

luacsoulet committed:
Add type For key: string array

luacsoulet committed:
update name

luacsoulet committed:
Enhance enemy spawn behavior and visual effects

- Add fade-in animation exclusively for swarm-type enemies
- Preserve formation shapes when spawning enemies outside screen boundaries
- Make saucer enemies appear from top of screen with faster entry
- Fix timer reset issues for enemy shooting patterns
- Optimize enemy movement with separate entry behavior
- Ensure formations maintain their relative positions during entry
- Reduce initial shooting delay for saucer enemies

luacsoulet committed:
Update MovementComponent.ts

luacsoulet committed:
Udpate some values for speed & fire rate

luacsoulet committed:
Add specialized formation spawning for saucer enemies

This commit enhances the FormationManager by implementing specialized behavior for saucer-type enemies in the default formation. When spawning saucer enemies:
- They are now placed at random horizontal positions across the screen width
- A margin of 50px is maintained from screen edges
- Vertical positions are slightly randomized (0-100px offset from base Y)
- This creates a more scattered, unpredictable pattern for saucer enemies
- Other enemy types in default formations continue to spawn at exact coordinates

luacsoulet committed:
Enhance MainGameScene with enemy spawning improvements and background parallax effects

- Added variable parallax scrolling for background elements
- Optimized enemy spawning logic with formation manager
- Implemented score display and registry tracking
- Added collision handling between player, enemies and bullets
- Set up preloading with progress bar for assets
